catgrad-core/src/interpreter/backend/candle.rs

```diff
@@ -0,0 +1,636 @@
+//
+// 1. **Trait Requirements**: The `NdArray<D>` trait requires `PartialEq` to be
+//    implemented for the tensor type. This is essential for:
+//    - Testing and debugging (comparing tensors in assertions)
+//    - Generic algorithms that need to compare tensor values
+//    - Hash-based data structures that require equality
+//
+// 2. **Backend Consistency**: Other backends (like ndarray) implement `PartialEq`
+//    naturally, so our Candle backend needs to match this interface for
+//    compatibility and interchangeability.
+//
+// 3. **Testing Infrastructure**: Many test frameworks and assertion macros
+//    rely on `PartialEq` to compare expected vs actual values. Without this,
+//    we cannot write meaningful tests for our tensor operations.
+//
+// 4. **API Ergonomics**: Users expect to be able to compare tensors using `==`
+//    and `!=` operators. This is a fundamental expectation in Rust APIs.
+//
+// IMPLEMENTATION CHOICES:
+//
+// Our current implementation compares only shape and dtype, not the actual values.
+// This is a deliberate trade-off for several reasons:
+//
+// 1. **Performance**: Comparing all tensor values would be expensive for large
+//    tensors, especially on GPU devices where data transfer is costly.
+//
+// 2. **Device Independence**: The underlying tensor data might be on different
+//    devices (CPU, GPU, Metal), making direct value comparison complex.
+//
+// 3. **Floating Point Precision**: Direct value comparison can be problematic
+//    with floating-point numbers due to precision issues and different
+//    computational paths.
+//
+// 4. **Use Case Alignment**: Most equality checks in the codebase are for
+//    structural equality (shape/dtype) rather than value equality.
+//
+// ALTERNATIVE APPROACHES:
+//
+// For cases where you need value-based equality, consider:
+// - `tensor.to_vec1()?` to extract values and compare manually
+// - Custom comparison functions with tolerance for floating-point values
+// - Specialized equality traits for different precision requirements
+//
+// This implementation strikes a balance between functionality and performance,
+// providing the necessary trait implementation while avoiding expensive operations.
+
+impl PartialEq for CandleBackend {
+    fn eq(&self, other: &Self) -> bool {
+        // Compare devices by their types only, since the underlying device IDs
+        // (CudaDevice, MetalDevice) don't implement PartialEq.
+        // For most use cases, this is sufficient as we typically care about
+        // whether we're using the same device type (CPU, CUDA, Metal) rather
+        // than specific device instances.
+        matches!(
+            (&self.device, &other.device),
+            (Device::Cpu, Device::Cpu)
+                | (Device::Cuda(_), Device::Cuda(_))
+                | (Device::Metal(_), Device::Metal(_))
+        )
+    }
+}
+
+impl PartialEq for CandleTensor {
```

Comment:
Requiring PartialEq for the NdArray type is a mistake in the design of the `Backend` trait. We should remove this - tracking in issue https://github.com/hellas-ai/catgrad/issues/190

This implementation seems fine for now, but we should remove once the constraint is gone.